design-proposals/klip-8-interactive-queries.md

```diff
@@ -131,13 +131,129 @@ Additionally:
+
+Currently, a streaming query against a table in KSQL will only return rows from the table's changelog produced *after* the query begins (assuming a default start offset of `latest`).
+
+However, streaming queries returning a table's entire materialized state *then* followed by changes is more consistent with the behavior of PIT queries on a table. This behavior is also likely what users want when issuing a streaming query against a table, and it is currently not straightforward to achieve.
+
+To get KSQL's current behavior with a streaming query against a table, users would simply use an intuitive `WHERE` clause:
+
+```sql
+SELECT * FROM table WHERE ROWTIME >= NOW EMIT CHANGES;
+```
+
+## Limitations
+
+The following are proposed intentional limitations of the initial implementation of PIT queries:
+
+**PIT queries on streams probably do not need to be initially supported.** The semantics and use cases around these are unclear and require further analysis. However, the proposed syntax leaves open the possibility of supporting PIT queries on streams without further modifications.
+
+**CTAS/CSAS defined by PIT queries do not initially need to be supported.** For example, the following CTAS would create a table, `t0`, and initially populate it with the result of the `SELECT` query:
+
+```sql
+CREATE TABLE t0 AS SELECT x, count(*) FROM t1 GROUP BY x;
+```
+
+**PIT queries against tables created directly over a topic (leaf tables) do not need to be initially supported.** Leaf tables are not materialized until another query reads from them. As a result, a PIT query against such a table would (potentially) not return the expected result. Rather than leave open the possibility of confusing query results in this scenario, we should simply disallow PIT queries against leaf tables until we decide to support them.
+
+---
+
+## Other considered approaches
+
+The following are other approaches that have been considered for identifying PIT versus streaming queries:
+
+### **Interpret any query against a table as PIT**
+
+The table-based approach would consider any SQL query against a KSQL table as a PIT query. Any query against a stream would then be a streaming query. The primary issue with this approach is that it complicates the capability of running a streaming query on a table, which is a highly valuable use case. Also, since PIT queries would be identified somewhat implicitly, this approach may not be particularly intuitive to users. It would also make it relatively easy to write a "correct" query that does the wrong thing.
+
+### **`SELECT STREAM[ING]`**
+
+Streaming queries would be identified by a new `STREAM[ING]` keyword: `SELECT STREAM[ING]...`. Similarly to `EMIT STREAM`, `SELECT STREAM[ING]` becomes somewhat redundant in the context of CSAS:
+
+```sql
+CREATE STREAM s0 AS SELECT STREAMING * FROM s1;
+```
+
+This approach also doesn't facilitate de-emphasizing the stream abstraction should we decide to eventually make KSQL more table centric.
```

Comment:
I read `make KSQL more table centric` the second time in this proposal and I have major concerns. Hence, I would propose to not use this idea to argue into any direction. The stream-table duality is the biggest strength of Kafka Streams and KSQL and I don't understand why people argue this direction.

Reply:
+1 . Stream/Table duality is lot clearer IMO, rather than calling everything a table at the SQL level, but having to explain its backed by a stream underneath.

Reply:
@mjsax @vinothchandar fair point, thank you both for chiming in. Another way to look at the underlying idea here is using constructs/semantics that are stream-table agnostic. This is the reason that I think `EMIT CHANGES` makes more sense for us versus `EMIT STREAM`: it's a single construct that conceptually makes sense on both tables and streams. In this way, stream-table agnostic constructs actually complement the stream-table duality quite nicely, IMO.

Reply:
I think you may be misinterpreting what 'more table centric' is meant to mean...

When we talk about stream/table duality we're more correctly talking about changelog/table duality.  That's not going anywhere - this is an important strength / feature of a steaming db. 

I believe when the Octo team are talking about making KSQL _more table centric_ they are referring to KSQL Stream vs KSQL Table.  In this sense a KSQL Stream, it is argued, is nothing more than a table without a primary key.

Having said that, I agree that we don't need this phrase to justify the work in this KLIP.